File: backend/app/schemas/snapshot.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import date, datetime
from decimal import Decimal


class SnapshotCreate(BaseModel):
    """Schema for creating a snapshot"""
    name: Optional[str] = None


class SnapshotEntryResponse(BaseModel):
    """Schema for snapshot entry response"""
    id: int
    category: str
    subcategory: Optional[str]
    name: str
    institution: Optional[str]
    account_type: Optional[str]
    amount: Decimal
    interest_rate: Optional[Decimal]

    class Config:
        from_attributes = True


class SnapshotGoalResponse(BaseModel):
    """Schema for snapshot goal response"""
    id: int
    goal_name: str
    target_amount: Decimal
    current_amount: Decimal
    target_date: Optional[date]
    completion_percentage: Optional[Decimal]

    class Config:
        from_attributes = True


class SnapshotResponse(BaseModel):
    """Schema for snapshot response"""
    id: int
    user_id: int
    snapshot_date: date
    snapshot_name: Optional[str]
    net_worth: Optional[Decimal]
    total_assets: Optional[Decimal]
    total_liabilities: Optional[Decimal]
    monthly_income: Optional[Decimal]
    monthly_expenses: Optional[Decimal]
    savings_rate: Optional[Decimal]
    age: Optional[int]
    employment_status: Optional[str]
    notes: Optional[str]
    created_at: datetime
    entries: List[SnapshotEntryResponse] = []
    goals: List[SnapshotGoalResponse] = []

    class Config:
        from_attributes = True


class TimelineData(BaseModel):
    """Schema for timeline chart data"""
    dates: List[str]
    net_worth: List[float]
    assets: List[float]
    liabilities: List[float]

File: frontend/src/types/snapshot.ts
export interface SnapshotEntry {
  id: number;
  category: string;
  subcategory?: string;
  name: string;
  institution?: string;
  account_type?: string;
  amount: number;
  interest_rate?: number;
}

export interface SnapshotGoal {
  id: number;
  goal_name: string;
  target_amount: number;
  current_amount: number;
  target_date?: string;
  completion_percentage?: number;
}

export interface Snapshot {
  id: number;
  user_id: number;
  snapshot_date: string;
  snapshot_name?: string;
  net_worth?: number;
  total_assets?: number;
  total_liabilities?: number;
  monthly_income?: number;
  monthly_expenses?: number;
  savings_rate?: number;
  age?: number;
  employment_status?: string;
  notes?: string;
  created_at: string;
  entries: SnapshotEntry[];
  goals: SnapshotGoal[];
}

export interface SnapshotCreate {
  name?: string;
}

export interface TimelineData {
  dates: string[];
  net_worth: number[];
  assets: number[];
  liabilities: number[];
}

export interface LastSnapshotInfo {
  last_snapshot_date?: string;
  days_since_last?: number;
  can_create_new: boolean;
}